lib/jdomodels/src/main/java/org/sagebionetworks/repo/model/dbo/file/download/v2/DownloadListDAOImpl.java

```diff
@@ -0,0 +1,170 @@
+import static org.sagebionetworks.repo.model.query.jdo.SqlConstants.COL_DOWNLOAD_LIST_2_PRINCIPAL_ID;
+import static org.sagebionetworks.repo.model.query.jdo.SqlConstants.COL_DOWNLOAD_LIST_2_UPDATED_ON;
+import static org.sagebionetworks.repo.model.query.jdo.SqlConstants.COL_DOWNLOAD_LIST_ITEM_2_ADDED_ON;
+import static org.sagebionetworks.repo.model.query.jdo.SqlConstants.COL_DOWNLOAD_LIST_ITEM_2_ENTITY_ID;
+import static org.sagebionetworks.repo.model.query.jdo.SqlConstants.COL_DOWNLOAD_LIST_ITEM_2_PRINCIPAL_ID;
+import static org.sagebionetworks.repo.model.query.jdo.SqlConstants.COL_DOWNLOAD_LIST_ITEM_2_VERION_NUMBER;
+import static org.sagebionetworks.repo.model.query.jdo.SqlConstants.TABLE_DOWNLOAD_LIST_2;
+import static org.sagebionetworks.repo.model.query.jdo.SqlConstants.TABLE_DOWNLOAD_LIST_ITEM_2;
+
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+import java.util.List;
+import java.util.stream.IntStream;
+
+import org.sagebionetworks.repo.model.download.ColumnName;
+import org.sagebionetworks.repo.model.download.DownloadListItem;
+import org.sagebionetworks.repo.model.download.IdAndVersion;
+import org.sagebionetworks.repo.model.download.SortDirection;
+import org.sagebionetworks.repo.model.jdo.KeyFactory;
+import org.sagebionetworks.repo.transactions.WriteTransaction;
+import org.sagebionetworks.util.ValidateArgument;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.jdbc.core.BatchPreparedStatementSetter;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.jdbc.core.RowMapper;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public class DownloadListDAOImpl implements DownloadListDAO {
+
+	public static final Long NULL_VERSION_NUMBER = -1L;
+
+	@Autowired
+	private JdbcTemplate jdbcTemplate;
+
+	private static final RowMapper<DBODownloadList> LIST_MAPPER = new DBODownloadList().getTableMapping();
+	private static final RowMapper<DBODownloadListItem> LIST_ITEM_MAPPER = new DBODownloadListItem().getTableMapping();
+
+	@WriteTransaction
+	@Override
+	public long addBatchOfFilesToDownloadList(Long userId, List<IdAndVersion> batchToAdd) {
+		ValidateArgument.required(userId, "User Id");
+		if (batchToAdd == null || batchToAdd.isEmpty()) {
+			return 0;
+		}
+		createOrUpdateDownloadList(userId);
+		IdAndVersion[] batchArray = batchToAdd.toArray(new IdAndVersion[batchToAdd.size()]);
+		int[] updates = jdbcTemplate.batchUpdate(
+				"INSERT IGNORE INTO " + TABLE_DOWNLOAD_LIST_ITEM_2 + " (" + COL_DOWNLOAD_LIST_ITEM_2_PRINCIPAL_ID + ","
+						+ COL_DOWNLOAD_LIST_ITEM_2_ENTITY_ID + "," + COL_DOWNLOAD_LIST_ITEM_2_VERION_NUMBER + ","
+						+ COL_DOWNLOAD_LIST_ITEM_2_ADDED_ON + ")  VALUES(?,?,?,NOW())",
+				new BatchPreparedStatementSetter() {
+
+					@Override
+					public void setValues(PreparedStatement ps, int i) throws SQLException {
+						IdAndVersion idAndVersion = batchArray[i];
+
+						if (idAndVersion == null) {
```

Comment:
Let add a test for this particular egde case (e.g. valid json array: [null, "123"]) and maybe improve the error message.